File: 0101-0200/0120.Triangle.js
//0ms

/**
 * @param {number[][]} triangle
 * @return {number}
 */
function minimumTotal(triangle) {
    const n = triangle.length;
    const dp = [...triangle[n - 1]]; // start with the last row

    // Move from second-last row up to the top
    for (let i = n - 2; i >= 0; i--) {
        for (let j = 0; j <= i; j++) {
            dp[j] = triangle[i][j] + Math.min(dp[j], dp[j + 1]);
        }
    }

    return dp[0];
}

File: 0101-0200/0120.Triangle.java
//0ms


class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {

        return helper(triangle, 0)[0];
    }

    int[] helper(List<List<Integer>> triangle, int p) {
        
        if(p == triangle.size()) {
            return new int[triangle.get(p-1).size()+1];
        }
        int down[] = helper(triangle, p+1);

        int ans[] = new int[triangle.get(p).size()];
        List<Integer> list = triangle.get(p);
        for(int i =0;i<ans.length;i++) {
            ans[i] = list.get(i)+Math.min(down[i], down[i+1]);
        }
        return ans;
    }
}
